### DISCUSSION TITLE
GitHub App ignoring additionalReviewers so not adding reviewers to PR

### DISCUSSION THREAD

**Discussion by nhojpatrick** [renovatebot/renovate#8421] [Request Help] [answered]
**What Renovate type, platform and version are you using?**

GitHub App

**Describe the bug**

Using this `"additionalReviewers" : ["nhojpatrick"],` in `.github/renovate.json` does not add `nhojpatrick` as a reviewer to the pull request.

**Relevant debug logs**

I can't see any error, but I do see;
```
DEBUG: Repository config
{
  "fileName": ".github/renovate.json",
  "config": {
    "extends": [
      "config:base"
    ],
    "dependencyDashboard": true,
    "excludeCommitPaths": [
      "relocation_**"
    ],
    "labels": [
      "dependencies"
    ],
    "additionalReviewers": [
      "nhojpatrick"
    ],
    "prHourlyLimit": 0,
    "separateMinorPatch": true,
    "separateMultipleMajor": true
  }
}
```

The below that I can see a branch being created and the pull request being created, the label is added but the reviewer section appears to be ignored.

```
DEBUG: Committing files to branch renovate/plugins.maven.maven-project-info-reports-plugin(branch="renovate/plugins.maven.maven-project-info-reports-plugin")
INFO: Branch created(branch="renovate/plugins.maven.maven-project-info-reports-plugin")
{
  "commitSha": "99de59a"
}
DEBUG: Ensuring PR(branch="renovate/plugins.maven.maven-project-info-reports-plugin")
DEBUG: There are 0 errors and 0 warnings(branch="renovate/plugins.maven.maven-project-info-reports-plugin")
DEBUG: getBranchPr(renovate/plugins.maven.maven-project-info-reports-plugin)(branch="renovate/plugins.maven.maven-project-info-reports-plugin")
DEBUG: findPr(renovate/plugins.maven.maven-project-info-reports-plugin, undefined, open)(branch="renovate/plugins.maven.maven-project-info-reports-plugin")
DEBUG: Creating PR(branch="renovate/plugins.maven.maven-project-info-reports-plugin")
{
  "prTitle": "Update dependency org.apache.maven.plugins:maven-project-info-reports-plugin to v3.1.1"
}
DEBUG: Creating PR(branch="renovate/plugins.maven.maven-project-info-reports-plugin")
{
  "title": "Update dependency org.apache.maven.plugins:maven-project-info-reports-plugin to v3.1.1",
  "head": "nhojpatrick:renovate/plugins.maven.maven-project-info-reports-plugin",
  "base": "develop",
  "draft": false
}
DEBUG: PR created(branch="renovate/plugins.maven.maven-project-info-reports-plugin")
{
  "pr": 24,
  "draft": false
}
DEBUG: Adding labels 'dependencies' to #24(branch="renovate/plugins.maven.maven-project-info-reports-plugin")
INFO: PR created(branch="renovate/plugins.maven.maven-project-info-reports-plugin")
{
  "pr": 24,
  "prTitle": "Update dependency org.apache.maven.plugins:maven-project-info-reports-plugin to v3.1.1"
}
DEBUG: Created Pull Request #24(branch="renovate/plugins.maven.maven-project-info-reports-plugin")
DEBUG: Checking #24 for automerge(branch="renovate/plugins.maven.maven-project-info-reports-plugin")
{
  "automerge": false,
  "automergeType": "pr",
  "automergeComment": "automergeComment"
}
DEBUG: No automerge(branch="renovate/plugins.maven.maven-project-info-reports-plugin")
```

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/master/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] This is a really small bug, it does not need a reproduction (think small typo)
- [ ] I have provided a minimal reproduction repository
- [x] I don't have time for that, but it happens in a public repository I have linked to
- [ ] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

Check any of my public repos, they all have renovate enabled, the same config and the same issue.

https://github.com/nhojpatrick/nhojpatrick-qa-ruleset-checkstyle/blob/develop/.github/renovate.json

**Comment by HonkingGoose** [comment]
I think you need to use the `reviewers` configuration option instead.

From the way I read the docs, the `additionalReviewers` config option is meant to add to the **existing** reviewers list.
You don't have any reviewers listed yet, so there's nothing to add.

It seems to me that `additionalReviewers` is meant to augment a preset or base list, without replacing a original `reviewers` list.
Quote from `additionalReviewers` docs:

> Additional reviewers for Pull Requests (in contrast to `reviewers`, this option adds to the existing reviewer list, rather than replacing it)

> In contrast to `reviewers`, this option adds to the existing reviewer list, rather than replacing it. This makes it suitable for augmenting a preset or base list without displacing the original, for example when adding focused reviewers for a specific package group.

---

Quote from `reviewers` docs:

> Requested reviewers for Pull Requests (either username or email address depending on the platform)
> Must be valid usernames. If on GitHub and assigning a team to review, use the prefix `team`:, e.g. provide a value like `team:someteam`.

---

I have opened a PR at your repository, with what I think is a better config (https://github.com/nhojpatrick/nhojpatrick-qa-ruleset-checkstyle/pull/30).

I recommend that you wait until the Renovate team has verified that my config is better, before you merge it in. 😄

**Comment by nhojpatrick** [comment]
@HonkingGoose yes I want to ensure a specifier reviewer is added to a PR. And I would quote the same documentation you highlighted.

Example
- Pull Request gets created and using `reviewer` abc is added.
- Total of 1 reviewer.
- User def is manually added.
- Total of 2 reviewers.
- Pull Request is now rebase.
- I would expect renovate to wipe out user def from the pull request, as `reviewers` is configured to be just `abc`.
- What is the rebase include an update to the the renovate configuration, should the new renovate configuration be used to update/overwrite the existing pull request.

Reading `additionalReviewers` I would expect it to work with `null`, `0` or `more than 1` reviewers.

If the documentation changes to state `1 or more` reviewers need to be defined before `additionalReviewers` is looked at, I would close this. I would then raise a feature request to add `ensureReviewers` that would work with `null`, `0` or `more than 1` reviewers where it ensures specified reviewers have been added to a pull request. Reading the previous issues opened which resulted in `additionalReviewers` being added, that is how I would expect it to work.

**Comment by rarkins** [comment]
Please re-test with `reviewers` and confirm if that fixes behaves as expected. If not then a minimal reproduction repo is necessary.

**Reply by adamko147** [reply]
I aggree with @nhojpatrick 
I had the same problem when using config presets from different repository. The preset from shared repo did not have reviewers, so no additionalReviewers added based on my config repo, which is quite confusing, as the documentation says

> In contrast to reviewers, this option adds to the existing reviewer list, rather than replacing it. This makes it suitable for augmenting a preset or base list without displacing the original, for example when adding focused reviewers for a specific package group.

Unfortunately I did not understand that empty list is not considered as _existing reviewer list_. I had to adjust the preset in shared repo, which is not exactly what I wanted - **I wanted to only add specific reviewer to one repo** without breaking any other existing presets

I've prepared minimal reproduce repo that simulates my case at https://github.com/adamko147/renovate-tests
- default.json is the preset, has some configuragion
- renovate.json is extending the "default.json" preset, adding additionalReviewers

I can confirm that additionalReviewers from renovate.json are added only if any "reviewer" is set in default.json



@rarkins Let me know what do you think. I've looked at the code and I can start work on changing the behaviour (remove the if `reviewers.length > 0` around https://github.com/renovatebot/renovate/blob/master/lib/workers/pr/index.ts#L69

**Reply by rarkins** [reply]
Yes, I think we should move this block outside that `if` check:

```
      if (config.additionalReviewers.length > 0) {
        const additionalReviewers = config.additionalReviewers.map(
          noLeadingAtSymbol
        );
        reviewers = [...new Set(reviewers.concat(additionalReviewers))];
      }
```

**Reply by adamko147** [reply]
Thanks, I will work on it and raise an PR.
@rarkins do you mind moving this discussion to issues (if that is possible) so I can refer to it in PR or should I rather create new issue and link this thread?
Thanks

**Reply by viceice** [reply]
You need to open a new issue, currently it's not possible to covert a discussion to an issue

**Reply by eana** [reply]
@rarkins I have both `assignees` and `reviewers` in the config file and still doesn't work. I will create reproduction repo later today.

**Reply by rarkins** [reply]
@eana please start a *new* discussion once you've done so, and if we confirm it's a bug and not config-related then we can create a bug issue after.

**Reply by eana** [reply]
@rarkins I tried to replicate the problem again and it turned out to be a config-related issue.

I totally missed this line in the log
```
DEBUG: Skipping assignees and reviewers as automerge=true(branch="renovate/docker-golangci-golangci-lint-1.x")
```

**Reply by nhojpatrick** [reply]
How i'm seeing this if you might not be in control of what you extend or have a corporate/company/team base which you are told to extend.

Under those condition you might not be aware of changes happening in the shared presets.

- So base doesn't have any `reviewers` set, so you need to use `reviewers`.
- Base might get changed to have `reviewers` and now your overwriting them without knowing you are.
- You spot that change and update usage from `reviewers` to `additionalReviewers`.
- Who ever maintains the base gets complains from the reviewers are now bombarded with reviews as they didn't know the volume they would get. So they change base again and remove `reviewers` again and now suddenly pr's have no reviewers.
- This is noticed again after a period of time pr's have no reviewers, so config has to change again from using `additionalReviewers` to `reviewers`

If `additionalReviewers` could cope with null, or empty existing `reviewers` it might be more helpful for those maintaining renovate config.

Not sure if any other fields might also be assisted with reviewing if they cope with null or empty from what they inherit.

**Reply by rarkins** [reply]
> If `additionalReviewers` could cope with null, or empty existing `reviewers` it might be more helpful for those maintaining renovate config.

I thought this was already fixed weeks ago by @adamko147. Have you tested it lately?

**Reply by nhojpatrick** [reply]
> > If `additionalReviewers` could cope with null, or empty existing `reviewers` it might be more helpful for those maintaining renovate config.
> 
> I thought this was already fixed weeks ago by @adamko147. Have you tested it lately?

Just retested 2021/03/12 on a GitHub project and reviewers set as per `additionalReviewers` field when extending `config:base`.

**Comment by nhojpatrick** [comment, marked as answer]
Use `additionalReviewers` to set reviewers as per documentation states. So you don't need to change config if `reviewers` is changed in the same renovate config or what you extend.
